crates/copilot/src/copilot_chat.rs

```diff
@@ -0,0 +1,260 @@
+            Self::Gpt3_5Turbo => "GPT-3.5",
+            Self::Gpt4 => "GPT-4",
+        }
+    }
+
+    pub fn max_token_count(&self) -> usize {
+        match self {
+            Self::Gpt4 => 8192,
+            Self::Gpt3_5Turbo => 16385,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Request {
+    pub intent: bool,
+    pub n: usize,
+    pub stream: bool,
+    pub temperature: f32,
+    pub model: Model,
+    pub messages: Vec<ChatMessage>,
+}
+
+impl Request {
+    pub fn new(model: Model, messages: Vec<ChatMessage>) -> Self {
+        Self {
+            intent: true,
+            n: 1,
+            stream: true,
+            temperature: 0.1,
+            model,
+            messages,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
+pub struct ChatMessage {
+    pub role: Role,
+    pub content: String,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub struct ResponseEvent {
+    pub choices: Vec<ResponseChoice>,
+    pub created: u64,
+    pub id: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ResponseChoice {
+    pub index: usize,
+    pub finish_reason: Option<String>,
+    pub delta: ResponseDelta,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ResponseDelta {
+    pub content: Option<String>,
+    pub role: Option<Role>,
+}
+
+pub async fn request_api_token(
+    oauth_token: &str,
+    api_url: &str,
+    client: Arc<dyn HttpClient>,
+    low_speed_timeout: Option<Duration>,
+) -> Result<(String, NaiveDateTime), anyhow::Error> {
+    let uri = format!("{api_url}/copilot_internal/v2/token");
+
+    let mut request_builder = HttpRequest::builder()
+        .method(Method::GET)
+        .uri(uri)
+        .header("Authorization", format!("token {}", oauth_token))
+        .header("Accept", "application/json");
+
+    if let Some(low_speed_timeout) = low_speed_timeout {
+        request_builder = request_builder.low_speed_timeout(100, low_speed_timeout);
+    }
+
+    let request = request_builder.body(AsyncBody::empty())?;
+
+    let mut response = client.send(request).await?;
+
+    if response.status().is_success() {
+        let mut body = Vec::new();
+        response.body_mut().read_to_end(&mut body).await?;
+
+        let body_str = std::str::from_utf8(&body)?;
+
+        let parsed: Value = serde_json::from_str(body_str)?;
+        Ok((
+            parsed["token"].as_str().unwrap().to_string(),
+            DateTime::from_timestamp(parsed["expires_at"].as_i64().unwrap(), 0)
+                .unwrap()
+                .naive_utc(),
+        ))
+    } else {
+        let mut body = Vec::new();
+        response.body_mut().read_to_end(&mut body).await?;
+
+        let body_str = std::str::from_utf8(&body)?;
+
+        Err(anyhow!("Failed to request API token: {}", body_str))
+    }
+}
+
+pub async fn fetch_copilot_oauth_token() -> Result<String, anyhow::Error> {
+    let path = if cfg!(target_os = "linux") {
```

Comment:
We should be using one of our well-known directory paths here instead of computing it on the fly: https://github.com/zed-industries/zed/blob/a5b8094082ec5c033e976572fb5cf48309e5d6b1/crates/paths/src/paths.rs

Reply:
I moved this logic into the paths package and simplified it a little bit at the same time. Let me know how it looks.